Exposed SHA3 through EVP API

havishr opened the pull request:
Changes: Implemented hash_evp_template.cpp.template, which exposes the EVP_MD API. Exposed SHA3_224, SHA3_256, SHA3_384, and SHA3_512. Allows for the use of these algorithms in an isolated context at the moment. 

Choices: Chose to use the specific constants described in the csrc files for each algorithm because they are not all exposed through EVP_MD. Manually copying the context from the Java state to EVP_MD_CTX and vice versa. Is there a better way of doing this? 

Testing: All testing was done the same as the other hash functions. Individual classes for each algorithm that contain KATs and the use HashFunctionTester. Did CAVP testing with the files from https://csrc.nist.gov/projects/cryptographic-algorithm-validation-program/secure-hashing. Had to set a minimum Java version for HashFunctionTester because SUN JDK8 does not expose SHA3 so trying to compare the implementations throws a NoSuchAlgorithmException. 


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

havishr committed:
Created hash_evp_template and added getContextSize, getHashSize, initContext, and updateContextByteArray with appropriate aws-lc functionality

havishr committed:
added all the function implementation for the template to hash_evp_template.cpp

havishr committed:
adjusted the build script for the hash spis to include the sha3 hashing. created the cpp for each hash with the appropriate definitions

havishr committed:
exposed the functionality of all four sha3 algos in accp. wrote the tests for sha3 excluding the HashFunctionTester implementations which need the appropriate zip files of data.

havishr committed:
Added test files and pointed to them in the new hash tests

havishr committed:
Reworked the init, both updates, and finish in the csrc code to work properly. Adjusted SPI names.

havishr committed:
Fixed bug in fast digest. All tests passing with regards to the SPIs.

havishr committed:
Adjusted variable definitions and shifted CTX definition back onto stack.

havishr committed:
Ran spotlessApply

havishr committed:
Undo .project file change

havishr committed:
Added checks to SHA3 tests to ensure Java Version is at least 11 because SHA3 is not exposed in SUN JDK 8 which is neede for TestAPI and TestRandomly

havishr committed:
Applied spotless

havishr committed:
Replaced duplicate constant

havishr committed:
Flattened repetitive testing logic for different SHA algorithms into one test class

havishr committed:
Applied spotless

havishr committed:
Adjusted to base testing class to maintain separate test classes while consolidating functionality.

havishr committed:
Ran spotless

WillChilds-Klein reviewed (commented):
Looking good so far!

havishr commented on csrc/hash_evp_template.cpp.template:
```diff
@@ -0,0 +1,243 @@
+#define TRANSFORM_sha3_256 SHA3256
+#define TRANSFORM_sha3_384 SHA3384
+#define TRANSFORM_sha3_512 SHA3512
+#define TRANSFORM_sha224 SHA224
+#define TRANSFORM_sha256 SHA256
+#define TRANSFORM_sha384 SHA384
+#define TRANSFORM_sha512 SHA512
+#define TRANSFORM_sha512_224 SHA512_224
+#define TRANSFORM_sha512_256 SHA512_256
+#define TRANSFORM_sha1 SHA1
+#define TRANSFORM_md5 MD5
+#define TRANSFORM_md4 MD4
+#define TRANSFORM_ripemd160 RIPEMD160
+#define TRANSFORM_shake128 SHAKE128
+#define TRANSFORM_shake256 SHAKE256
+#define TRANSFORM_blake2b256 BLAKE2B256
```
Its there for future functionality if needed, but it isn't part of this implementation, so I can remove it.

havishr committed:
Fixed up changes related to comments including unnecessary files and cleaning up files

havishr committed:
Fixed up comments related to cleaning up certain files

WillChilds-Klein commented on csrc/hash_evp_template.cpp.template:
```diff
@@ -0,0 +1,207 @@
+
+        java_buffer contextBuffer = java_buffer::from_array(env, contextArray);
+        {            
+            CHECK_OPENSSL(EVP_DigestInit(&ctx, MD()));
```
same question about whether we need this.

havishr committed:
Added comments to clarify buffer usage.

WillChilds-Klein committed:
Address PR #485 review feedback

- Fix md_data leak: use scoped EVP_MD_CTX (EVP_MD_CTX_auto) in
  initContext/updateContextByteArray/updateNativeByteBuffer so the
  EVP_DigestInit-allocated md_data is freed on every call.
- Move the JNI class-name token (JAVA_CLASS_NAME) into each
  per-algorithm .cpp so the shared template is algorithm-agnostic.
- Consolidate MD_CTX_SIZE into the template and document it and
  DIGEST_BLOCK_SIZE (no public API exposes the internal ctx size).
- Explain java_buffer vs bounce_buffer usage; clean up whitespace and
  the fast-path scratch comment.
- Restore the unrelated .project deletion.
- Refactor BaseSHATest to take algorithm/pre-image/vectors/CAVP files
  via constructor; fold duplicate CAVP tests into one file-list test;
  inline the SHA*Test constants.
